File: Strings/Easy/ValidParentheses.c
#include<stdio.h>
#include<string.h>
int main(){
  char str[100];
  fgets(str,100,stdin);
  str[strcspn(str,"\n")]='\0';
  char stack[100];
  int top=-1;
  int flag=1;
  for(int i=0;str[i]!='\0';i++){
    char c = str[i];
    if(c=='('||c=='['||c=='{'){
      stack[++top]=str[i];
    }
    else if(c==')'||c==']'||c=='}'){
        if(top==-1){
          flag=0;
          break;
        }
        else if((c==')'&&stack[top]!='(')
              ||(c==']'&&stack[top]!='[')
              ||(c=='}'&&stack[top]!='{')){
                flag=0;
                break;
              }
        top--;
      }
    }
    if(top!=-1)flag=0;
    if(flag)printf("valid");
    else printf("Not valid");
}

File: Strings/Easy/ValidParentheses.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String str = sc.nextLine();
        char[] stack = new char[str.length()];
        int top = -1;
        boolean valid = true;
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if (c == '(' || c == '[' || c == '{') {
                stack[++top] = c;
            } 
            else if (c == ')' || c == ']' || c == '}') {
                if (top == -1) {
                    valid = false;
                    break;
                }
                char open = stack[top--];
                if ((c == ')' && open != '(') ||
                    (c == ']' && open != '[') ||
                    (c == '}' && open != '{')) {
                    valid = false;
                    break;
                }
            }
        }
        if (top != -1)
            valid = false;

        System.out.println(valid ? "Valid" : "Invalid");
    }
}
